File: examples/php/behavioral/state.php
<?php

interface PhoneState
{
    public function pickUp(): PhoneState;
    public function hangUp(): PhoneState;
    public function dial(): PhoneState;
}

class PhoneStateIdle implements PhoneState
{
    public function pickUp(): PhoneState {
        return new PhoneStatePickedUp();
    }
    public function hangUp(): PhoneState {
        throw new Exception("already idle");
    }
    public function dial(): PhoneState {
        throw new Exception("unable to dial in idle state");
    }
}

class PhoneStatePickedUp implements PhoneState
{
    public function pickUp(): PhoneState {
        throw new Exception("already picked up");
    }
    public function hangUp(): PhoneState {
        return new PhoneStateIdle();
    }
    public function dial(): PhoneState {
        return new PhoneStateCalling();
    }
}

class PhoneStateCalling implements PhoneState
{
    public function pickUp(): PhoneState {
        throw new Exception("already picked up");
    }
    public function hangUp(): PhoneState {
        return new PhoneStateIdle();
    }
    public function dial(): PhoneState {
        throw new Exception("already dialing");
    }
}

class Phone {
    private $state;

    public function __construct() {
        $this->state = new PhoneStateIdle();
    }
    public function pickUp() {
        $this->state = $this->state->pickUp();
    }
    public function hangUp() {
        $this->state = $this->state->hangUp();
    }
    public function dial() {
        $this->state = $this->state->dial();
    }
}

$phone = new Phone();

$phone->pickUp();
$phone->dial();

File: examples/python/behavioral/state.py
class Phone:
    def __init__(self):
        self.state = IdleState()

    def pick_up(self):
        self.state.pick_up(self)

    def hang_up(self):
        self.state.hang_up(self)


class IdleState:
    def pick_up(self, phone):
        print("Phone picked up")
        phone.state = CallingState()

    def hang_up(self, phone):
        print("Phone is already idle")


class CallingState:
    def pick_up(self, phone):
        print("Already calling")

    def hang_up(self, phone):
        print("Call ended")
        phone.state = IdleState()


phone = Phone()
phone.pick_up()
phone.hang_up()
